### PR TITLE
Download pre-compiled bootstrap from CI

### LINKED ISSUE #99989: Download a pre-compiled bootstrap from CI
Currently, we unconditionally recompile bootstrap on all builds, like libstd. In most cases, this is unnecessary: bootstrap is rarely changed and it's confusing that we have to build a whole rust program before processing `x.py --help`. Additionally, it means that when we add the new shell script entrypoints, we'd have to replicate all of bootstrap.py's logic for finding config.toml and the build directory. If we can download bootstrap through rustup, it will greatly simplify all that logic.

cc @dvtkrlbs

Mentoring instructions:
1. Add a new `download_or_build_bootstrap` function to `src/bootstrap/bootstrap.py`.
2. If there are changes to src/bootstrap (you can find these with `git diff-index --quiet HEAD -- src/bootstrap`), then always rebuild it from source.
3. Otherwise, download it from ci-artifacts.rust-lang.org, as configured by `stage0.json` - you may have to modify the `Stage0Toolchain` python object to load this configuration.

Helps with https://github.com/rust-lang/rust/issues/94829.

### PR DESCRIPTION
**Pull request by dvtkrlbs**
This PR builds on the work done by #98483 and downloads the bootstrap from the artifacts server if there is not any changes on the bootstrap source code. If the download fails it fallbacks to local build. 

fixes #99989 
helps with #94829

### CODE REVIEW

**Comment by dvtkrlbs** [comment]
i could not find the button to turn the pr into draft it right now works but it is not a clean solution. I am not sure how to clean this it has a few hacks on a few functions to change behaviour

**Comment by dvtkrlbs** [comment]
RIght now the wip code still tries to download even if there is a bootstrap locally. I am not sure what is the best way to handle this case. I first thought about chekcing if the cache  folder has a bootstrap binary and then skipping download but this does not consider the case when user updates their old copy of the source code and in this case x.py will still use old bootstrap binary. What do you think is the best way to handle this case ?

**Comment by jyn514** [comment]
i don't have time for reviews, sorry. onur is working on the shell script side, he's a good reviewer.

r? @ozkanonur

**Comment by ozkanonur** [comment]
I can do the review by the coming weekend. Until then, you can squash your commits and undraft the PR once it's ready.

@rustbot author

**Comment by dvtkrlbs** [comment]
I think this is ready for initial review the redownload logic is not fully clean but I am open to suggestions for best way to detect changes ?

File: src/bootstrap/bootstrap.py [hunk 1]
```diff
@@ -1057,7 +1076,25 @@ def bootstrap(args):
     # Fetch/build the bootstrap
     build.download_toolchain()
     sys.stdout.flush()
-    build.build_bootstrap(args.color, args.warnings, verbose_count)
+
+
+    try:
+        if not did_path_changed("src/bootstrap"):
```
Comment (ozkanonur) [inline src/bootstrap/bootstrap.py:1082]: this path will not work when `x` called from outside of the rust tree.

**Review by onur-ozkan — CHANGES_REQUESTED** [review]
Maybe it's time to update this log

https://github.com/rust-lang/rust/blob/1673e54f679ebf67fc7c889b4bcba6d93add0456/src/bootstrap/bootstrap.py#L1124

Something like `Checking if bootstrap needs to be downloaded or built before processing.` should fit better

File: src/bootstrap/bootstrap.py [hunk 2]
```diff
@@ -1057,7 +1076,25 @@ def bootstrap(args):
...
+            print(commit_hash)
+            filename = "{}-{}.tar.xz".format("bootstrap-nightly", build.build_triple())
+            success = build._download_component_helper(commit_hash, filename, "bootstrap", "tar.xz")
+            if success:
+                build.is_precompiled_bootstrap = True
+    except:
+        print("warn: downloading bootstrap from ci failed falling back to local build")
+
+    if not build.is_precompiled_bootstrap:
+        build.build_bootstrap(args.color, args.warnings, verbose_count)
```
[the remark this answers is not in the document]
Comment (dvtkrlbs) [reply src/bootstrap/bootstrap.py:1095]: The one problem with this kind of flow is I think we should fallback to local build if download fails should I just invoke build bootstrap on the except branch as well ?

[the remark this answers is not in the document]
Comment (ozkanonur) [reply src/bootstrap/bootstrap.py:1095]: > I just invoke build bootstrap on the except branch as well ?

Except might be caused from building the bootstrap, so it's better to be handled under `try`. I was trying to give general idea(cutting off the download logic and making it look more simpler)

File: src/bootstrap/bootstrap.py (hunk 1 above)
Comment (dvtkrlbs) [reply src/bootstrap/bootstrap.py:1082]: What could be a better way to check path changes I think there was some work done by you in a PR about bootstrap-shim binary but I could not find the PR?

**Comment by dvtkrlbs** [comment]
The one problem that I did not find a clean solution for still is even if there is a built bootstrap binary in the build folder and there is no changes it still tries to download so I think the `bootstrap_modified` also needs to detect if there is an up to date bootstrap binary. I am not sure how to best solve this.

File: src/bootstrap/bootstrap.py (hunk 1 above)
Comment (ozkanonur) [reply src/bootstrap/bootstrap.py]: > What could be a better way to check path changes

Checking with git is completely fine, but `src/bootstrap` is a relative path and will not work when there is no `src/bootstrap` in the current working directory. There are many people(including me) using `x` from outside of the rust tree.

> I think there was some work done by you in a PR about bootstrap-shim binary but I could not find the PR?

Here https://github.com/rust-lang/rust/pull/107812/files#diff-1e2a2bd13194bee1cb25532bfe2cd41f9d7bf25ca62f42f32ea302d97de71356R55-R79

**Comment by ozkanonur** [comment]
> I think the bootstrap_modified also needs to detect if there is an up to date bootstrap binary. I am not sure how to best solve this.

Yes, this should be checked too. You can create a stamp file on the same path with bootstrap binary and check that for this particular check.

File: src/bootstrap/bootstrap.py (hunk 1 above)
Comment (dvtkrlbs) [reply src/bootstrap/bootstrap.py]: > Here [#107812 (files)](https://github.com/rust-lang/rust/pull/107812/files#diff-1e2a2bd13194bee1cb25532bfe2cd41f9d7bf25ca62f42f32ea302d97de71356R55-R79)

where is config.last_modified_commit defined? The config.rs file doesn't seem to contain the function. I also did a code search and still couldn't find the function.

Comment (ozkanonur) [reply src/bootstrap/bootstrap.py]: > where is config.last_modified_commit defined? The config.rs file doesn't seem to contain the function. I also did a code search and still couldn't find the function.

https://github.com/ozkanonur/rust/blob/8ad89068e2558940b4184d46763ff9c6b325883b/src/bootstrap/min_config.rs#L178-L233

**Comment by dvtkrlbs** [comment]
I added the stamp check.

File: src/bootstrap/bootstrap.py (hunk 1 above)
Comment (dvtkrlbs) [reply src/bootstrap/bootstrap.py]: I changed the logic to first get the toplevel and also started passing cwd to the Popen.

File: src/bootstrap/bootstrap.py [hunk 3]
```diff
@@ -574,20 +586,59 @@ def download_toolchain(self):
...
+                verify_checksum=False,
             )
+            except:
+                return False
         unpack(tarball, tarball_suffix, self.bin_root(), match=pattern, verbose=self.verbose)
+        return True
+
+
+
+    def download_bootstrap(self, commit_hash):
```
Comment (ozkanonur) [inline src/bootstrap/bootstrap.py:611]: Too many spaces

File: src/bootstrap/bootstrap.py [hunk 4]
```diff
@@ -574,20 +586,59 @@ def download_toolchain(self):
...
+
+        path = "{}/src/bootstrap".format(top_level)
+        cmd = ["git", "diff-index", "--quiet", "HEAD", "--", path]
+
+        ret = subprocess.Popen(cmd, cwd=self.rust_root)
+        code = ret.wait()
+        if code != 0:
+            return None
+        else:
+            return merge_base
```
Comment (ozkanonur) [inline src/bootstrap/bootstrap.py:632]: This command alone `git log -1 --pretty=format:%H src/bootstrap` does what you need

Comment (ozkanonur) [reply src/bootstrap/bootstrap.py:632]: if you need short version of the commit hash, simply lowercase the `H` in `--pretty`

File: src/bootstrap/bootstrap.py [hunk 5]
```diff
@@ -574,20 +586,59 @@ def download_toolchain(self):
...
+
+    def bootstrap_out_of_date(self, commit):
+        stamp_path = pathlib.Path(self.bin_root()).joinpath("bootstrap/.bootstrap-stamp")
+        if not stamp_path.exists():
+            return True
+
+        return stamp_path.read_text != commit
```
Comment (ozkanonur) [inline src/bootstrap/bootstrap.py:639]: This seems wrong as it will always return `True` because we don't have the stamp file on fresh repository.

**Comment by ozkanonur** [comment]
I haven't looked into this in detail, but I think the following steps should get the job done easily:

- Check if anything has been changed in the src/bootstrap folder. If there are any modifications, compile the bootstrap.

- If there haven't been any changes, first check if we have a stamp file:
- - If we do, compare its value with the latest bootstrap commit. If they don't match, download the latest version from the CI system.
- - If we don't have a stamp file, download it from the CI system.

Regardless of whether we need to compile or download, we need to make sure the stamp file has been generated.

**Comment by dvtkrlbs** [comment]
Wait do we distribute the stamp-file with the CI as well. I thought it gets created after a successful download to mark the artifact.

**Comment by ozkanonur** [comment]
> Wait do we distribute the stamp-file with the CI as well.

No, this is not needed.

**Comment by dvtkrlbs** [comment]
```
fmt check
tidy check
dependency exception `dunce` license has changed
    previously `CC0-1.0 OR MIT-0 OR Apache-2.0` now `CC0-1.0 OR MIT-0`
    update EXCEPTIONS for the new license
some tidy checks failed
Build completed unsuccessfully in 0:00:37
error: failed to push some refs to 'github.com:dvtkrlbs/rust'
```
trying to rebasae my branch and getting this error. and it is not helpful at all it first does not tell which crate has this dependency and second it does not tell where this exceptions file or sections is

**Comment by dvtkrlbs** [comment]
oh found it it was in the tools tidy deps

**Comment by Kobzol** [comment]
https://github.com/rust-lang/rust/blob/214a0d1531cc74454b1af0a379b74436d73aa614/src/tools/tidy/src/deps.rs#L57

**Comment by Dylan-DPC** [comment]
@dvtkrlbs any updates on resolving the conflicts?

**Comment by dvtkrlbs** [comment]
I will update my branch in the following days. Was there an easy way to resolve submodule conflicts there was a way to force reset them to HEAD without going through the conflict resolving but I don't remember how.

**Pull request closed without merging**